demo/api/src/products/product-persist.pipe.ts

```diff
@@ -0,0 +1,130 @@
+import { BlockDataInterface } from "@comet/blocks-api";
+import { DamImageBlock } from "@comet/cms-api";
+import { Connection, EntityManager, FilterQuery, IDatabaseDriver, Reference } from "@mikro-orm/core";
+import { LoggerService } from "@nestjs/common";
+import { ImporterPipe, PipeMetadata } from "@src/importer/pipes/importer-pipe.type";
+import { Transform, TransformCallback } from "stream";
+import { v4 } from "uuid";
+
+import { Product } from "./entities/product.entity";
+import { ProductCategory } from "./entities/product-category.entity";
+import { ProductColor } from "./entities/product-color.entity";
+import { ProductImporterInput } from "./product-importer.input";
+
+type ProductData = Omit<ProductImporterInput, "image"> & {
+    id: string;
+    colors: string;
+    category: Reference<ProductCategory>;
+    image: BlockDataInterface;
+};
+
+export class ProductPersistPipe implements ImporterPipe {
+    constructor(private readonly em: EntityManager<IDatabaseDriver<Connection>>) {}
+
+    getPipe(logger: LoggerService) {
+        return new ProductPersist(this.em, logger);
+    }
+}
+
+class ProductPersist extends Transform {
+    private persistedEntitiesAmount: number;
+
+    constructor(private readonly em: EntityManager<IDatabaseDriver<Connection>>, private readonly logger: LoggerService) {
+        super({ writableObjectMode: true, objectMode: true });
+        this.persistedEntitiesAmount = 0;
+    }
+
+    getPersistedEntitiesAmount() {
+        return this.persistedEntitiesAmount;
+    }
+
+    async _transform(inputDataAndMetadata: { data: ProductData; metadata: PipeMetadata }, encoding: BufferEncoding, callback: TransformCallback) {
+        try {
+            const { data } = inputDataAndMetadata;
+            const { colors: colorString, category, colors: colorCollection, ...productData } = data;
+
+            const updateQuery: FilterQuery<object> | undefined = { slug: data.slug };
+            const record: Product | null = updateQuery ? await this.em.findOne(Product, updateQuery, { fields: ["id"] }) : null;
+
+            productData.image = DamImageBlock.blockInputFactory({
+                activeType: "pixelImage",
+                attachedBlocks: [
+                    {
+                        props: {},
+                        type: "pixelImage",
+                    },
+                ],
+            }).transformToBlockData();
+
+            // TODO: use upsert
+            // const product = await this.em.upsert(Product, productInput, {
+            //     onConflictFields: ["slug"],
+            //     onConflictAction: "merge",
+            //     onConflictExcludeFields: ["id", "createdAt"],
+            // });
```

Comment:
Do you plan to fix this in an upcoming PR? Or this one?

Reply:
In an upcoming PR. There is a problem with the ImageBlock. It does not work with insert atm. We have to look for a solution separately.